src/frontend/src/optimizer/rule/stream/stream_filter_expression_simplify_rule.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use fixedbitset::FixedBitSet;
+use risingwave_common::types::ScalarImpl;
+use risingwave_connector::source::DataType;
+
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, FunctionCall};
+use crate::optimizer::plan_expr_visitor::strong::Strong;
+use crate::optimizer::plan_node::{ExprRewritable, LogicalFilter, LogicalShare, PlanTreeNodeUnary};
+use crate::optimizer::rule::{BoxedRule, Rule};
+use crate::optimizer::PlanRef;
+
+/// Specially for the `StreamFilter` under `StreamShare`
+pub struct StreamFilterExpressionSimplifyRule {}
+impl Rule for StreamFilterExpressionSimplifyRule {
+    /// The pattern we aim to optimize, e.g.,
+    /// 1. (NOT (e)) OR (e) => True
+    /// 2. (NOT (e)) AND (e) => False
+    /// NOTE: `e` should only contain at most a single column
+    /// otherwise we will not conduct the optimization
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let filter: &LogicalFilter = plan.as_logical_filter()?;
+        let mut rewriter = StreamFilterExpressionSimplifyRewriter {};
+        let logical_share_plan = filter.input();
+        let share: &LogicalShare = logical_share_plan.as_logical_share()?;
+        let input = share.input().rewrite_exprs(&mut rewriter);
+        share.replace_input(input);
+        Some(LogicalFilter::create(
+            share.clone().into(),
+            filter.predicate().clone(),
+        ))
+    }
+}
```

Comment:
Can we change this rule to a `logical` rule instead of `streaming` only.

Reply:
This indeed could be applied to batch plan as well, what I'm not sure is that, will `BatchShare` construct the same `BatchFilter` with this kind of rare optimizable pattern? Just like I said here: https://github.com/risingwavelabs/risingwave/pull/15275#issuecomment-1965683819.

And regarding the simplification of `BatchFilter` predicate, [this PR](https://github.com/risingwavelabs/risingwave/pull/15699) aims to resolve this. (especially under `LogicalScan`)

Reply:
I change it to more general rewrite for `LogicalFilter`, so planner tests will change a lot accordingly.